### DISCUSSION TITLE
How to declare a function with more than one statement?

### DISCUSSION THREAD

**Discussion by 0x70b1a5** [josdejong/mathjs#3128] [Q&A] [answered]
I'm trying to collect several columns of mathjs statements and execute them each against an incrementing counter `x` (1, 2, 3...). After each column has been executed by mathjs, extract the value of `y`.

For example: 

```
System input:
x=0    x=1    x=2    x=3
------------------------
User input:
a=1    a=4    a=x        
y=a    y=a-4  y=a-1  y=x
------------------------
Output: 
1      0      1      3
```

There is an additional complication to my problem: each column is 1 second, and we desire to create the waveform of an audio file. That means, between each column, interpolate the values of `x` 44099 times (for sample rate 44100).

It's easy enough to create a range for x, `x*44100:1:(x+1)*44100`. But the tricky part is how to evaluate each of the intervening columns to efficiently extract `y`. You can see how this might be necessary for a sine wave which fluctuates many times a second, with something like `y=sin(x)`. We don't just want the audio file to have a point at `x=0` seconds and another point at `x=1` seconds with nothing in between.

To do this efficiently, my first idea was to declare the entire column as an `f(x)`, eval it with mathjs, and extract `y`. But it seems like mathjs functions can only consist of a single statement.

```
Reading column 1:
f(x) = x=0; a=1; y=a; y
Output:
1
```

For example, this fails:

```
f(x) = d = 1; d
f(3) // expect 1; actual: error
```

I understand this is the classic "XY problem" (ha!), but I will certainly accept an answer that solves my X problem instead.

**Comment by dvd101x** [comment, marked as answer]
I think there are specific plans to implement functions with multiple statements according to #2701

In the meantime a workaround is that elements of an array can include expressions with assignments.

```jl
f(x) = [d = 1, d][end] # evaluate all expressions in the array and return only the value of the last expression.
f(3)                   # 1
```
You have to be very careful with your naming convention as the expressions will assign values in the scope (there are not function scopes)

```jl
d = 5
f(x) = [d = 1, d][end] # evaluate all expressions in the array and return only the value of the last expression.
d                      # 5
f(3)                   # yields 1 but overrides the value of d
d                      # 1
```

Now for your original problem, I don't think that functions with multiple steps are mandatory for your problem. Maybe it's more of an issue with vectorization, can you use an additional dimension to solve your issue in one go?

**Reply by 0x70b1a5** [reply]
Ah, that's brilliant! Thank you.

I'm not sure exactly how to use vectorization to solve the issue differently. What do you mean?

**Reply by dvd101x** [reply]
I'm glad it was useful!

What I mean by vectorization is: Think of X as a vector, A as a vector and Y as a vector

So you could do your original system as:
```jl
X = [0, 1, 2, 3]
A = [1,  4,  X[3], 0]
C = [0, -4, -1,   X[4]]
Y = A + C                 # [1, 0, 1, 3]
```
Including this logic as a single function
```jl
f(X) = [1, 4, X[3], 0] + [0, -4, -1, X[4]]
f([0, 1, 2, 3])           # [1, 0, 1, 3]
```
If you need to repeat this many times, you can do that in an additional dimension or as much as needed. So maybe there is not even the need for a function just plain matrix addition as mathjs does broadcasting.

Here are a few interesting reads:
- https://www.mathworks.com/help/matlab/matlab_prog/vectorization.html
- https://numpy.org/doc/stable/user/basics.broadcasting.html

**Reply by 0x70b1a5** [reply]
Ah, thank you for explaining!

"A" was just an example equation in my post, and is meaningless to the program logic. We expect users to be able to declare their own variables in mathjs syntax, write new functions, etc, all within one of the columns. This is part of why I was thinking of having each column be its own function...

**Reply by dvd101x** [reply]
Ok, then maybe your needs are more in line with a function that receives functions:

```jl
A(X) = [1, 4, X[3], 0];
C(X) = [0, -4, -1, X[4]];
Y(X, A) = A(X) + C(X);
f(X, A, Y) = Y(X, A);

X = [0, 1, 2, 3];
f(X, A, Y)          # [1, 0, 1, 3]
```

**Comment by gwhitney** [comment]
See also #3355 for another suggestion of how to do this at least partially
